File: src/features/emergency-contacts/services/contactService.ts
import { supabase } from "@/core/database/client";
import { AppError, ValidationError } from "@/shared/errors";
import type {
  AdminContactSummary,
  ContactDirectoryDraft,
  ContactDirectoryEntry,
} from "../domain/types";
import { toDomain, toInsert } from "./mappers";

const requireUserId = async (): Promise<string> => {
  const { data } = await supabase.auth.getUser();
  if (!data.user) throw new AppError("Not authenticated", "AUTH_REQUIRED");
  return data.user.id;
};

const validate = (draft: ContactDirectoryDraft) => {
  if (!draft.name?.trim())
    throw new ValidationError("Name is required", "name");
  if (!draft.phone?.trim())
    throw new ValidationError("Phone is required", "phone");
};

export const contactService = {
  async list(): Promise<ContactDirectoryEntry[]> {
    const { data, error } = await supabase
      .from("emergency_contacts")
      .select("*")
      .order("priority", { ascending: true });
    if (error) throw new AppError(error.message, "CONTACTS_LIST_FAILED", error);
    return (data ?? []).map(toDomain);
  },

  async create(draft: ContactDirectoryDraft): Promise<ContactDirectoryEntry> {
    validate(draft);
    const userId = await requireUserId();
    const { data, error } = await supabase
      .from("emergency_contacts")
      .insert(toInsert(userId, draft))
      .select()
      .single();
    if (error)
      throw new AppError(error.message, "CONTACTS_CREATE_FAILED", error);
    return toDomain(data);
  },

  async remove(id: string): Promise<void> {
    const { error } = await supabase
      .from("emergency_contacts")
      .delete()
      .eq("id", id);
    if (error)
      throw new AppError(error.message, "CONTACTS_DELETE_FAILED", error);
  },

  /**
   * Admin/moderation view: every user's emergency contacts, newest first,
   * joined with a minimal user profile summary for display/search.
   */
  async listAllForAdmin(): Promise<AdminContactSummary[]> {
    const { data, error } = await supabase
      .from("emergency_contacts")
      .select(
        `
        id,
        name,
        phone,
        email,
        relationship,
        priority,
        user_id,
        created_at
      `,
      )
      .order("created_at", { ascending: false });
    if (error) throw new AppError(error.message, "CONTACTS_LIST_FAILED", error);

    const rows = data ?? [];
    const userIds = [...new Set(rows.map((r) => r.user_id))];
    const { data: profiles, error: profilesError } = await supabase
      .from("profiles")
      .select("id, full_name, phone_number")
      .in("id", userIds);
    if (profilesError)
      throw new AppError(profilesError.message, "CONTACTS_LIST_FAILED", profilesError);

    const profileMap = (profiles ?? []).reduce(
      (acc, profile) => {
        acc[profile.id] = profile;
        return acc;
      },
      {} as Record<string, { id: string; full_name: string | null; phone_number: string | null }>,
    );

    return rows.map((row) => {
      const profile = profileMap[row.user_id];
      const entry = toDomain(row as unknown as Parameters<typeof toDomain>[0]);
      return {
        ...entry,
        userProfile: {
          fullName: profile?.full_name ?? "Unknown User",
          phoneNumber: profile?.phone_number ?? "Not provided",
        },
      };
    });
  },
};

File: mobile/lib/services/contact_service.dart
import '../core/supabase_client.dart';
import '../models/activity_entry.dart';
import '../models/emergency_contact.dart';
import 'activity_log_service.dart';

/// Wired to the real `emergency_contacts` table (RLS-scoped to the signed-in
/// user via `user_id`). Same shape as the original's
/// ContactDirectoryEntry/ContactDirectoryDraft types.
class ContactService {
  ContactService._();
  static final ContactService instance = ContactService._();

  ContactRelationship? _relationshipFromLabel(String? label) {
    if (label == null) return null;
    for (final r in ContactRelationship.values) {
      if (r.label == label) return r;
    }
    return null;
  }

  EmergencyContact _fromRow(Map<String, dynamic> row) {
    return EmergencyContact(
      id: row['id'] as String,
      name: row['name'] as String,
      phone: row['phone'] as String,
      email: row['email'] as String?,
      relationship: _relationshipFromLabel(row['relationship'] as String?),
      priority: (row['priority'] as int?) ?? 1,
      createdAt: DateTime.parse(row['created_at'] as String),
    );
  }

  Future<List<EmergencyContact>> list() async {
    final userId = supabase.auth.currentUser?.id;
    if (userId == null) return [];
    final rows = await supabase
        .from('emergency_contacts')
        .select()
        .eq('user_id', userId)
        .order('priority', ascending: true);
    return (rows as List).map((r) => _fromRow(r as Map<String, dynamic>)).toList();
  }

  Future<EmergencyContact> create({
    required String name,
    required String phone,
    String? email,
    ContactRelationship? relationship,
    int priority = 1,
  }) async {
    final userId = supabase.auth.currentUser?.id;
    if (userId == null) throw Exception('You must be signed in to add a contact.');
    if (name.trim().isEmpty || phone.trim().isEmpty) {
      throw Exception('Please fill in at least the name and phone number.');
    }
    final row = await supabase
        .from('emergency_contacts')
        .insert({
          'user_id': userId,
          'name': name.trim(),
          'phone': phone.trim(),
          'email': (email?.trim().isEmpty ?? true) ? null : email!.trim(),
          'relationship': relationship?.label,
          'priority': priority,
        })
        .select()
        .single();
    final contact = _fromRow(row);
    ActivityLogService.instance.log(ActivityType.contactChange, 'Added contact: ${contact.name}');
    return contact;
  }

  Future<void> update(EmergencyContact updated) async {
    await supabase.from('emergency_contacts').update({
      'name': updated.name,
      'phone': updated.phone,
      'email': updated.email,
      'relationship': updated.relationship?.label,
      'priority': updated.priority,
    }).eq('id', updated.id);
  }

  Future<void> remove(String id) async {
    await supabase.from('emergency_contacts').delete().eq('id', id);
  }
}
